base_layer/core/src/validation/block_validators.rs

```diff
@@ -296,19 +296,9 @@ fn check_inputs_are_utxos<B: BlockchainBackend>(block: &Block, db: &B) -> Result
 }
 
 // This function checks that the inputs and outputs do not exist in the STxO set.
-fn check_not_stxos<B: BlockchainBackend>(block: &Block, db: &B) -> Result<(), ValidationError> {
-    for input in block.body.inputs() {
```

Comment:
While it's true that in the current order of validators, this check is unnecessary, I don't think this shouldn't be removed. Instead it should be called before `check_inputs_are_utxos`. The reason is that the block may be put into different places depending on the validation result. Example: If a block contains a known double spend, then it is clearly invalid and should be discarded. If the UTXO in the input does not exist, it may be put in the orphan pool.

Reply:
Currently if any of them fail it results in the same location. Double spend and an unknown input result in orphan pool. 
The chance of having the same tx in an orphan pool with a chain split, is close 100%. In this case its the same tx and its not really an invalid block. Its just invalid for this chain. 

All of these checks are for the statefull validator, which checks on the current chain. Orphan pool results are already checked by now so we know these blocks might be valid. If any of these checks fail, we know the blocks are not valid for the current chain.

Reply:
There are two stages to validation: the orphan validators (if validation fails the block will not enter the orphan pool) and the "full" block validator which comes after. A failure here will result in the block not being entered into the main chain (however it is already in the orphan pool at this point).
 
A double spend within a block is checked at the orphan stage by `check_duplicate_transactions_inputs` - if that fails it will not enter the orphan pool. The "inputs are stxos" check happens on the "main chain" stage but I think it's redundant because the database implementations ensure that a given commitment cannot be in two spend states at once, so the UTXO check is sufficient. 

Any optimisation here are a boon for block sync times.